log4j-core/src/main/java/org/apache/logging/log4j/core/impl/ThreadContextDataInjector.java

```diff
@@ -183,11 +183,13 @@ public StringMap injectContextData(final List<Property> props, final StringMap r
             // When the ThreadContext is garbage-free, we must copy its key-value pairs into the specified reusable
             // StringMap. We cannot return the ThreadContext's internal data structure because it may be modified later
             // and such modifications should not be reflected in the log event.
-            copyProperties(props, reusable);
+            final StringMap actualReusable = reusable != null && !reusable.isFrozen() ? reusable
+                    : ContextDataFactory.createContextData();
```

Comment:
Doesn't this imply that, unless certain conditions are met, `ForGarbageFreeThreadContextMap.injectContextData()` is not garbage-free anymore.

Reply:
Yes, you are right. Since `AsyncLoggerConfig` is the one making an immutable copy of the `RingBufferLogEvent`, it should bear the responsibility for the allocation.

BTW: the logic behind copying mutable/immutable objects should be reviewed in `3.x`. E.g. under certain conditions `AsyncLoggerConfig` uses mutable events in its ring buffer. In such a case there is no need make a copy of the thread context: a mutable destination copy already exists.